tests/AulaTests/Stories/Interpreter/Action.hs

```diff
@@ -69,52 +82,119 @@ runClient (Free (SelectIdeaSpace s k)) = do
         Left v'  -> Action.markIdeaInJuryPhase (idea ^. _Id) v'
         Right v' -> Action.markIdeaInResultPhase (idea ^. _Id) v'
+    postcondition $ do
+        Just idea' <- findIdeaByTitle t
+        case v of
+            Left  v' -> (idea' ^? ideaJuryResult . _Just . ideaJuryResultValue) `shouldBe` Just v'
+            Right v' -> (idea' ^? ideaVoteResult . _Just . ideaVoteResultValue) `shouldBe` Just v'
     runClient k
 
 runClient (Free (VoteIdea t v k)) = do
-    Just idea <- findIdeaByTitle t
-    _ <- lift $ Action.voteIdea (idea ^. _Id) v
+    Just idea <- precondition $ findIdeaByTitle t
+    _ <- step . lift $ Action.voteIdea (idea ^. _Id) v
+    postcondition $ do
+        Just idea' <- findIdeaByTitle t
+        let noOfVotes  = Map.size $ idea  ^. ideaVotes
+        let noOfVotes' = Map.size $ idea' ^. ideaVotes
+        -- FIXME: The same user can vote only once
+        noOfVotes' `shouldBe` (noOfVotes + 1)
+    runClient k
+
+runClient (Free (CommentIdea t c k)) = do
+    Just idea <- precondition $ findIdeaByTitle t
+    _ <- step . lift $
+        currentUserAddDb (AddCommentToIdea (idea ^. _Id)) (Markdown c)
+    postcondition $ checkIdeaComment t c
     runClient k
 
+runClient (Free (CommentOnComment t cp c k)) = do
+    (idea, comment) <- precondition $ do
+        Just idea <- findIdeaByTitle t
+        let Just comment = findCommentByText idea cp
+        return (idea, comment)
+    _ <- step . lift $
+        currentUserAddDb
+            (AddReplyToIdeaComment (idea ^. _Id) (comment ^. _Id))
+            (Markdown c)
+    postcondition $ checkIdeaComment t c
+    runClient k
+
+  where
+    checkIdeaComment t c = do
+        Just idea' <- findIdeaByTitle t
+        let Just _comment = findCommentByText idea' c
+        return ()
+
 -- * helpers
 
-findIdeaByTitle :: (ActionM m) => IdeaTitle -> StateT ClientState m (Maybe Idea)
+findIdeaByTitle :: (ActionM m) => IdeaTitle -> ActionClient m (Maybe Idea)
 findIdeaByTitle t = fmap (find ((t ==) . view ideaTitle)) . lift $ query getIdeas
 
-findTopicByTitle :: (ActionM m) => IdeaTitle -> StateT ClientState m (Maybe Topic)
+findTopicByTitle :: (ActionM m) => IdeaTitle -> ActionClient m (Maybe Topic)
 findTopicByTitle t = fmap (find ((t ==) . view topicTitle)) . lift $ query getTopics
 
-assert :: (Show msg, ActionM m) => msg -> Bool -> m ()
+findCommentByText :: Idea -> CommentText -> Maybe Comment
+findCommentByText i t = find ((t ==) . fromMarkdown . _commentText) . Map.elems $ i ^. ideaComments
+
+assert :: (Show msg, Monad m) => msg -> Bool -> m ()
 assert _ True  = return ()
 assert msg False = error $ "assertion failed: " <> show msg
     -- FIXME: give source code location of the call.
+
+shouldBe :: (Monad m, Eq a, Show a) => a -> a -> m ()
+shouldBe actual expected = assert (actual, expected) (actual == expected)
+    -- FIXME: give source code location of the call.
+
+-- ** Notations for test step sections
+
+precondition :: Monad m => m a -> m a
```

Comment:
Follows the BDD convention, `given`, `when`, `then` combined with the contract based approach `precondition`, `step`, `postcondition`. At the moment these combinators has no effects at all. The other approach to this separation are add comments all over the place `-- precondtion` etc, which I would like to avoid.